### PR TITLE
fix: issues on WHERE search filter

### PR DESCRIPTION
**Pull request by techchintan**
Close: Point 3: - Filter - https://github.com/SigNoz/signoz/issues/2353#issuecomment-1518713817
Except for Subpoint - 4 & 6

### CODE REVIEW

File: frontend/src/api/queryBuilder/getAttributesValues.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	attributeKey,
+	attributeKeyDataType,
+	filterAttributeTagType,
+	searchText,
+}: IGetAttributeValuesPayload): Promise<
+	SuccessResponse<IAttributeValuesResponse> | ErrorResponse
+> => {
+	try {
+		const response = await ApiV3Instance.get(
+			`/autocomplete/attribute_values?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&filterAttributeTagType=${filterAttributeTagType}&searchText=${searchText}&attributeKey=${attributeKey}&attributeKeyDataType=${attributeKeyDataType}`,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getAttributesValues.ts:23]: can we leverage `lib/createQueryParams` ?

File: frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx [hunk 2]
```diff
@@ -73,6 +79,11 @@ function QueryBuilderSearch({
 		return tags;
 	}, [query.aggregateAttribute.key, tags]);
 
+	const isMatricsDataSource = useMemo(
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx:82]: this does not require useMemo

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 3]
```diff
@@ -55,48 +59,52 @@ export const useFetchKeysAndValues = (
...
+		}
+
+		const { payload } = await getAttributesValues({
+			aggregateOperator: query.aggregateOperator,
+			dataSource: query.dataSource,
+			aggregateAttribute: query.aggregateAttribute.key,
+			attributeKey,
+			attributeKeyDataType: query.aggregateAttribute.dataType,
+			filterAttributeTagType: query.aggregateAttribute.type ?? null,
+			searchText: !result[result.length - 1]?.endsWith(',')
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:79]: can you write small comment for searchText why we need this one

**Review by palashgdev — COMMENTED** [review]
small change... though working fine from functionality perspective

File: frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx (hunk 2 above)
Comment (techchintan) [reply frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx:82]: @palashgdev - It is required because `query.datasource` is a mutable prop.

**Comment by srikanthccv** [comment]
@techchintan please incorporate the changes from this commit https://github.com/SigNoz/signoz/commit/b91a13ed4d4b2471644d27eef506b57a9b96c194. 

And I noticed the value of the query params is incorrect for the `/attribute_values` API. Here is what each of them means. This was working earlier.

- `aggregateOperator`: aggregation operator selected from the list of aggregations supported for each signal.
- `dataSource`: one of the data sources selected from logs, traces, and metrics.
- `aggregateAttribute`: aggregation attribute key selected
- `attributeKey`: The key from the where clause the user has typed. (where keys are fetched from `/attribute_keys` API)
- `tagType`: tag type of the key from the where clause the user has typed. (`/attribute_keys` would have also provided the tag type, that should be used here, null in case of a custom key)
- `filterAttributeKeyDataType`: data type of the key from the where clause user has typed. (`/attribute_keys` would have also provided the data type, that should be used here, null in case of a custom key)
- `searchText`: search text for the value

**Comment by srikanthccv** [comment]
I do not think the first point is addressed probably related to incorrect query param. The key-op-value flow is not starting unless the aggregate attribute is chosen.

**Comment by srikanthccv** [comment]
https://www.loom.com/share/8505b9807b544c25ba765d91ddd7129c

File: frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx [hunk 4]
```diff
@@ -109,8 +120,8 @@ function QueryBuilderSearch({
 			notFoundContent={isFetching ? <Spin size="small" /> : null}
 		>
 			{options?.map((option) => (
-				<Select.Option key={option.value} value={option.value}>
-					{option.value}
+				<Select.Option key={option.value} value={option.label}>
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx:123]: don't you think this will break while using value ?

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 5]
```diff
@@ -25,25 +25,42 @@ type UseFetchKeysAndValuesReturnValues = {
 export const useFetchKeysAndValues = (
 	searchValue: string,
 	query: IBuilderQueryForm,
-): UseFetchKeysAndValuesReturnValues => {
-	const [keys, setKeys] = useState<AttributeKeyOptions[]>([]);
+): IuseFetchKeysAndValues => {
+	const [keys, setKeys] = useState<BaseAutocompleteData[]>([]);
 	const [results, setResults] = useState<string[]>([]);
+
+	const isQueryEnabled = useMemo(
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:32]: @techchintan no need of memo as return value is primitive

File: frontend/src/hooks/queryBuilder/useOptions.ts [hunk 6]
```diff
@@ -18,33 +19,47 @@ export const useOptions = (
...
-					value: `${key} ${o}`,
-					label: `${key} ${o.replace('_', ' ')}`,
+					value: `${key} ${o} `,
+					label: `${key} ${o.replace('_', ' ')} `,
 				})),
 			);
 		} else if (key && operator) {
 			if (isMulti) {
-				setOptions(results.map((r) => ({ value: `${r}` })));
+				setOptions(results.map((r) => ({ label: `${r}`, value: `${r}` })));
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useOptions.ts:51]:
```suggestion
				setOptions(results.map((option) => ({ label: `${option}`, value: `${option}` })));
```

File: frontend/src/hooks/queryBuilder/useOptions.ts [hunk 7]
```diff
@@ -18,33 +19,47 @@ export const useOptions = (
...
+			condition: !data?.isColumn,
+		});
+
 	const updateOptions = useCallback(() => {
 		if (!key) {
 			setOptions(
 				searchValue
-					? [{ value: searchValue }, ...keys.map((k) => ({ value: k.key }))]
-					: keys?.map((k) => ({ value: k.key })),
+					? [
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useOptions.ts:33]: make helper function for this one... don't dump in one place

File: frontend/src/hooks/queryBuilder/useOptions.ts [hunk 8]
```diff
@@ -18,33 +19,47 @@ export const useOptions = (
 	const [options, setOptions] = useState<Option[]>([]);
 	const operators = useOperators(key, keys);
 
+	const getLabel = (data: BaseAutocompleteData): Option['label'] =>
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useOptions.ts:22]: memo

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx (hunk 4 above)
Comment (techchintan) [reply frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/index.tsx:123]: We are using Label not value @palashgdev.

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts (hunk 5 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:32]: This is not a primitive value case, the left side of the condition is a non-primitive  `query.dataSource`
So based on your `query.dataSource` it will change boolean.

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 9]
```diff
@@ -54,49 +76,56 @@ export const useFetchKeysAndValues = (
...
-					} else {
-						setResults([]);
-					}
-				}
-			}
+		if (!value) {
+			return;
+		}
+		const { tagKey, tagOperator, tagValue } = getTagToken(value);
+		const filterAttributeKey = keys.find((key) => tagKey.includes(key.key));
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:85]: Why is this `tagKey.includes(key.key)`?

Say I have keys as
```javascript
keys = [
    {
        "key": "http_status",
        "dataType": "number",
        "type": "tag",
        "isColumn": false
    },
    {
        "key": "http_status_text",
        "dataType": "string",
        "type": "tag",
        "isColumn": false
    }
]

```

and tagKey = "http_status_text", why would this send `http_status` as `attributeKey` and `number` as `filterAttributeKeyDataType ` when it should be sending `http_status_text` and `string`?

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 10]
```diff
@@ -54,49 +76,56 @@ export const useFetchKeysAndValues = (
...
-						setResults(values);
-					} else {
-						setResults([]);
-					}
-				}
-			}
+		if (!value) {
+			return;
+		}
+		const { tagKey, tagOperator, tagValue } = getTagToken(value);
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:84]: Since this is using the `TAG_FSM`, there is no need to mandate the user to add space, or there is no need for us to add some custom logic to add space.

File: frontend/src/hooks/queryBuilder/useSetCurrentKeyAndOperator.ts [hunk 11]
```diff
@@ -1,27 +1,27 @@
...
+			const { tagKey, tagOperator, tagValue } = getTagToken(value);
+			const isSuggestKey = keys?.some((el) => tagKey.includes(el.key));
 
 			if (getCountOfSpace(value) >= 1 || isSuggestKey) {
-				key = tKey || '';
-				operator = tOperator || '';
-				result = tResult.filter((el) => el);
+				key = tagKey || '';
+				operator = tagOperator || '';
+				result = tagValue || [];
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useSetCurrentKeyAndOperator.ts:24]: What is happening here? What is `getCountOfSpace` or `isSuggestKey` doing?

File: frontend/src/hooks/queryBuilder/useTagValidation.ts [hunk 12]
```diff
@@ -19,7 +19,9 @@ export const useTagValidation = (
 	result: string[],
 ): ITagValidation => {
 	const operatorType = useOperatorType(operator);
-	const isValidTag = useIsValidTag(operatorType, result.length);
+	const resultLength =
+		operatorType === 'SINGLE_VALUE' ? [result]?.length : result?.length;
+	const isValidTag = useIsValidTag(operatorType, resultLength);
 
 	const { isExist, isValidOperator, isMulti, isFreeText } = useMemo(() => {
 		const isExist = operatorType === QUERY_BUILDER_SEARCH_VALUES.NON;
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useTagValidation.ts:27]: There is `checkStringEndsWithSpace`. Why is that needed? What does it decide?

File: frontend/src/hooks/queryBuilder/useTag.ts [hunk 13]
```diff
@@ -47,9 +49,10 @@ export const useTag = (
 			) {
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useTag.ts:49]: Just above this line, there is `isExistsNotExistsOperator(value)`. I am constantly surprised by this kind of logic. If anyone thinks about a moment, it's easy to think where it breaks. This logic allows something like this. No operator no value.
Screenshot 2023-05-06 at 4 23 50 PM

File: frontend/src/hooks/queryBuilder/useTag.ts [hunk 14]
```diff
@@ -61,13 +64,14 @@ export const useTag = (
...
-			(query?.tagFilters?.items || []).map((obj) =>
-				isInNotInOperator(obj.op)
-					? `${obj.key} ${obj.op} ${obj.value.join(',')}`
-					: `${obj.key} ${obj.op} ${obj.value.join(' ')}`,
-			),
+		const initialTags = (query?.filters?.items || []).map((obj) =>
+			isInNInOperator(getOperatorFromValue(obj.op))
+				? `${obj.key?.key} ${getOperatorFromValue(
+						obj.op,
+				  )} ${(obj.value as string[]).join(', ')}`
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useTag.ts:71]: Among other issues, the commas are only a small part, and here is such an issue. The naive join breaks. say the value is `["a", "b,c", "d"]`, the `.join` will create `'a, b,c, d'` which in turn will create a value as `['a', ' b', 'c', ' d']` which is incorrect. You should use `unparse` with `{data: value}`. When you write a code, take a moment to think about what your doing and how it breaks.

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 15]
```diff
@@ -49,18 +64,18 @@ export const useAutoComplete = (query: IBuilderQuery): IAutoComplete => {
...
-						return prev.replace(` ${value}`, '');
-					}
-					return checkStringEndsWithSpace(prev)
-						? `${prev} ${value}`
-						: `${prev} ${value},`;
+					const matches = prev?.matchAll(TAG_FSM);
+					const [match] = matches ? Array.from(matches) : [];
+					const [, , , matchTagValue] = match;
+					const data = Papa.parse(matchTagValue).data.flat();
+					return replaceStringWithMaxLength(prev, data as string[], value);
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:78]: I see you got rid of some white space and replaced it with `matchAll` from the expression. What does this code do?

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 16]
```diff
@@ -130,5 +145,7 @@ export const useAutoComplete = (query: IBuilderQuery): IAutoComplete => {
 		searchValue,
 		isMulti,
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:146]: There is some `handleKeyDown` logic above. It's not clear what that is doing. If I use the arrow key up-down for selecting a value for `IN`, it instead makes a chip. It doesn't deselect but makes a chip. There must be some way to handle this. How do others do this?

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts (hunk 9 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts]: @srikanthccv 
https://github.com/SigNoz/signoz/pull/2629/commits/879051d28b579927c8087bbda08584327655ae7f

File: frontend/src/hooks/queryBuilder/useTag.ts (hunk 13 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useTag.ts:49]: @srikanthccv 
https://github.com/SigNoz/signoz/pull/2629/commits/172bc0fb2892e1e3e82db79de0ddba85fc9f2d93

File: frontend/src/hooks/queryBuilder/useTagValidation.ts (hunk 12 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useTagValidation.ts:25]: @srikanthccv - Earlier It was used for IN/NOT_IN for space around options you select. but not it's no more used. As Space for IN/NOT_IN is mandatory.
https://github.com/SigNoz/signoz/pull/2629/commits/a7ba777ca05063683fb38cc44dbbda2316e16d51

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts (hunk 15 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useAutoComplete.ts:75]: This code will replace the last search string with the selected option or the user will allow a custom key. @srikanthccv 
The previous value will return all options you selected(string with comma separator) + at the ends with your searched value.
hence `replaceStringWithMaxLength` will do the same.

File: frontend/src/hooks/queryBuilder/useTag.ts (hunk 13 above)
Comment (srikanthccv) [reply frontend/src/hooks/queryBuilder/useTag.ts]: That's still not a correct way to fix it. `isExistsNotExistsOperator` should only return true when the operator is `EXISTS` or `NOT_EXISTS` not the simple text search.

File: frontend/src/hooks/queryBuilder/useTag.ts (hunk 14 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useTag.ts]: @srikanthccv - Removed join from value
https://github.com/SigNoz/signoz/pull/2629/commits/7517cbf499a5d2e9f22add2b1ac77d30f2bd90bf

File: frontend/src/hooks/queryBuilder/useSetCurrentKeyAndOperator.ts (hunk 11 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useSetCurrentKeyAndOperator.ts:24]: @srikanthccv - Removed space logic
https://github.com/techchintan/signoz/commit/ec381985ecc9812e83e0cca23990d3543f57b341

File: frontend/src/hooks/queryBuilder/useTag.ts [hunk 17]
```diff
@@ -61,13 +61,14 @@ export const useTag = (
...
-		setTags(
-			(query?.tagFilters?.items || []).map((obj) =>
-				isInNotInOperator(obj.op)
-					? `${obj.key} ${obj.op} ${obj.value.join(',')}`
-					: `${obj.key} ${obj.op} ${obj.value.join(' ')}`,
-			),
-		);
+		const initialTags = (query?.filters?.items || []).map((ele) => {
+			if (isInNInOperator(getOperatorFromValue(ele.op))) {
+				const csvString = Papa.unparse([ele.value]);
```
Comment (srikanthccv) [inline frontend/src/hooks/queryBuilder/useTag.ts:66]: This doesn't work. This is not the right way to use the `unparse`.

File: frontend/src/hooks/queryBuilder/useOptions.ts [hunk 18]
```diff
@@ -18,37 +20,63 @@ export const useOptions = (
...
+					: getOptionsFromKeys(keys),
 			);
 		} else if (key && !operator) {
 			setOptions(
-				operators?.map((o) => ({
-					value: `${key} ${o}`,
-					label: `${key} ${o.replace('_', ' ')}`,
+				operators?.map((operator) => ({
+					value: `${key}${operator} `,
+					label: `${key}${operator} `,
```
[the remark this answers is not in the document]
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useOptions.ts]: @srikanthccv https://github.com/SigNoz/signoz/pull/2629/commits/06f0b77e25108b66a14432b4c35489c64b135cf0

File: frontend/src/hooks/queryBuilder/useTag.ts (hunk 17 above)
Comment (techchintan) [reply frontend/src/hooks/queryBuilder/useTag.ts:66]: @srikanthccv @palashgdev 
https://github.com/SigNoz/signoz/pull/2629/commits/7e1dc7076f147c2b9a22ca990a7f492a9ca9b0b0

File: frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/utils.ts [hunk 19]
```diff
@@ -102,3 +102,7 @@ export function replaceStringWithMaxLength(
 	if (lastSearchValue === '') return `${mainString}${replacementString},`; // if user select direclty from options
 	return mainString.replace(lastSearchValue, `${replacementString},`);
 }
+
+export function checkCommaInValue(str: string): string {
+	return str.includes(',') ? `"${str}"` : str;
+}
```
Comment (srikanthccv) [inline frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/utils.ts]: Please do not use the basic string functions. Please avoid doing that for non-trivial tasks. Why can't you use unparse here?

Comment (techchintan) [reply frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/utils.ts]: This is just more of a utility function, it's just checking whether the string(Value) has a comma or not. @srikanthccv 
CC: @techchintan

Comment (techchintan) [reply frontend/src/container/QueryBuilder/filters/QueryBuilderSearch/utils.ts]: @srikanthccv @palashgdev 
input: ['**chintan**','**HTTP GET,GET,PUT**','**Sudani,POST**','**Super,"luxurious",truck**']
output: 
image

Is this scenario right?

**Review by palashgdev — APPROVED** [review]
will be taking the feedback in another as functionality is working fine
